packages/rollup-contracts/contracts/CanonicalTransactionChain.sol

```diff
@@ -0,0 +1,115 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+/* Internal Imports */
+import {DataTypes as dt} from "./DataTypes.sol";
+import {RollupMerkleUtils} from "./RollupMerkleUtils.sol";
+import {L1ToL2TransactionQueue} from "./L1ToL2TransactionQueue.sol";
+
+contract CanonicalTransactionChain {
+  address public sequencer;
+  uint public sequencerLivenessAssumption;
+  RollupMerkleUtils public merkleUtils;
+  L1ToL2TransactionQueue public l1ToL2Queue;
+  uint public cumulativeNumElements;
+  bytes32[] public batches;
+  uint public lastOVMTimestamp;
+
+  constructor(
+    address _rollupMerkleUtilsAddress,
+    address _sequencer,
+    address _l1ToL2TransactionPasserAddress,
+    uint _sequencerLivenessAssumption
+  ) public {
+    merkleUtils = RollupMerkleUtils(_rollupMerkleUtilsAddress);
+    sequencer = _sequencer;
+    l1ToL2Queue = new L1ToL2TransactionQueue(_rollupMerkleUtilsAddress, _l1ToL2TransactionPasserAddress, address(this));
+    sequencerLivenessAssumption =_sequencerLivenessAssumption;
+    lastOVMTimestamp = 0;
+  }
+
+  function getBatchesLength() public view returns (uint) {
+    return batches.length;
+  }
+
+  function hashBatchHeader(
+    dt.TxChainBatchHeader memory _batchHeader
+  ) public pure returns (bytes32) {
+    return keccak256(abi.encodePacked(
+      _batchHeader.timestamp,
+      _batchHeader.isL1ToL2Tx,
+      _batchHeader.elementsMerkleRoot,
+      _batchHeader.numElementsInBatch,
+      _batchHeader.cumulativePrevElements
+    ));
+  }
+
+  function authenticateAppend(address _sender) public view returns (bool) {
+    return _sender == sequencer;
+  }
+
+  function appendL1ToL2Batch() public {
+    dt.TimestampedHash memory timestampedHash = l1ToL2Queue.peek();
+    uint timestamp = timestampedHash.timestamp;
+    if (timestamp + sequencerLivenessAssumption > now) {
+      require(authenticateAppend(msg.sender), "Message sender does not have permission to append this batch");
+    }
+    lastOVMTimestamp = timestamp;
+    bytes32 elementsMerkleRoot = timestampedHash.txHash;
+    uint numElementsInBatch = 1;
+    bytes32 batchHeaderHash = keccak256(abi.encodePacked(
+      timestamp,
+      true, // isL1ToL2Tx
+      elementsMerkleRoot,
+      numElementsInBatch,
+      cumulativeNumElements // cumulativePrevElements
+    ));
+    batches.push(batchHeaderHash);
+    cumulativeNumElements += numElementsInBatch;
```

Comment:
Noting that this seems like a reasonable place for an `appendBatch(time, isL1, numElements)` function, seems like this functionality would be shared in a few places